### ISSUE TITLE
Unable to install extensions from a company mirror

### ISSUE THREAD

**Issue by bnk2** [mozilla/policy-templates#412]
Installing official AddOns from a company internal mirror fails.

Occurs with the following OS and Firefox versions
* Linux (Debian 10)
  * Firefox 68.0.1 (x64)
* Windows 10 (1903)
  * Firefox 68.0.1esr (x64)

It **does** work referring the AddOns to the official mirror ```https://addons.mozilla.org``` with a firefox instance with access to the *Internet*.
It does **not** when referring another internal webserver from a firefox instance without access to the *Internet*

```about:policies#errors``` says
> Download failed - https://some.company-host.net/some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi
> Download failed - https://some.company-host.net/some/path/to/mirrored/addons/german_dictionary-2.0.6.1webext.xpi?

Investigating the download attempt with the developer tool inside Firefox (networking) - and also trying with *curl* or *wget* - works fine and responds with ```HTTP/2.0 200 OK```.

Response headers for the *English dictionary* file:
```
HTTP/2.0 200 OK
server: nginx/1.10.3
date: Wed, 24 Jul 2019 11:11:45 GMT
content-type: application/x-xpinstall
content-length: 214136
last-modified: Thu, 17 Jan 2019 16:04:44 GMT
etag: "5c40a79c-34478"
strict-transport-security: max-age=31536000
x-frame-options: sameorigin
x-xxs-protection: 1; mode=block
x-content-type-options: nosniff
accept-ranges: bytes
X-Firefox-Spdy: h2
```
It also works downloading the extensions (with another HTTP client) and installing the AddOns as a local file by using the ```about:addons``` dialog.

The following policy (excerpt) is used (and loaded - all other customization are working fine):
```javascript
{
  "AppUpdateURL": "",
  "policies": {
    "BlockAboutAddons": false,
    ...
    "Certificates": {
      "ImportEnterpriseRoots": true,
      "Install": [
        "C:\\Temp\\SOME_COMPANY_CERT.crt",
        "/usr/local/share/ca-certificates/SOME_COMPANY_CERT.crt"
      ]
    },
    ...
    "DefaultDownloadDirectory": "${home}/Downloads",
    "DisableAppUpdate": true,
    ...
    "DisableSecurityBypass": {
      "InvalidCertificate": true,
      "SafeBrowsing": true
    },
    ...
    "DisableSystemAddonUpdate": false,
    "Extensions": {
      "Install": [
        "https://some.company-host.net/some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi",
        "https://some.company-host.net/some/path/to/mirrored/addons/german_dictionary-2.0.6.1webext.xpi"
      ]
    },
    "ExtensionUpdate": true,
    ...
    "InstallAddonsPermission": {
      "Allow": [
        "https://some.company-host.net/"
      ],
      "Default": true
    },
    "NetworkPrediction": false,
    ...
    "PopupBlocking": {
      "Locked": false,
      "Default": true
    },
    ...
    "PromptForDownloadLocation": false,
    ...
    "SSLVersionMax": "tls1.3",
    "SSLVersionMin": "tls1.2"
  }
}
```

Please provide some further help.

**Comment by mkaply** [comment]
Weird. The error coming through is this one:

https://searchfox.org/mozilla-central/source/browser/components/enterprisepolicies/Policies.jsm#1585

From the addons manager. I'll do some local downloads to see if I can recreate.

I'll also have us produce a better error in that case.

**Comment by mkaply** [comment]
I've tested some local installs and things are working for me. I'm going to add a better error message in so I can see what is going on. In the mean time, we can try to debug this. 

Here's how you can do it. Go to about:config. Search for devtools.chrome.enabled. Set it to true.

Go to the web developer menu in the main browser menu and open the scratchpad. In the top level menu, change the environment for the scratchpad from content to browser. Then paste this in:

installAddonFromURL("PUT YOUR URL HERE");

function installAddonFromURL(url, extensionID) {
  AddonManager.getInstallForURL(url, {
    telemetryInfo: { source: "enterprise-policy" },
  }).then(install => {
    if (install.addon && install.addon.appDisabled) {
      alert(`Incompatible add-on - ${location}`);
      install.cancel();
      return;
    }
    let listener = {
      /* eslint-disable-next-line no-shadow */
      onDownloadEnded: install => {
        if (extensionID && install.addon.id != extensionID) {
          alert(
            `Add-on downloaded from ${url} had unexpected id (got ${
              install.addon.id
            } expected ${extensionID})`
          );
          install.removeListener(listener);
          install.cancel();
        }
        if (install.addon && install.addon.appDisabled) {
          alert(`Incompatible add-on - ${url}`);
          install.removeListener(listener);
          install.cancel();
        }
      },
      onDownloadFailed: () => {
        install.removeListener(listener);
        alert(`Download failed - ${install.error} - ${url}`);
      },
      onInstallFailed: () => {
        install.removeListener(listener);
        alert(`Installation failed - ${url}`);
      },
      onInstallEnded: () => {
        install.removeListener(listener);
        alert(`Installation succeeded - ${url}`);
      },
    };
    install.addListener(listener);
    install.install();
  });
}


There should be a number in the error popup (-1, -2, -3)

I would need that number to help debug,.

**Comment by bnk2** [comment]
The result a failed Download with error code `-1`.
> Download failed - -1 - https://some.company-host.net/some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi

I just tried changing the Protocol/Scheme of the download URL to `http` and surprisingly the result was...
> Installation succeeded - http://some.company-host.net/some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi

The webserver is configured to redirect HTTP-requests with response code `302` to HTTPS. HTTP Protocol version `2.0` is configured for HTTPS requests only.

```javascript
server {
        listen 80;
        listen [::]:80;
        server_name some.company-host.net;
        ...
        location / {
                ...
                return 302 https://$host$request_uri;
        }
        ...
}
server {
        listen 443 ssl http2;
        listen [::]:443 ssl http2;
        server_name some.company-host.net;
        ...
}
```

The HTTP protocol version is is the only major difference between the HTTP response headers of our internal mirror and the Mozilla AddOn mirror.

Maybe the issue is related to the HTTP protocol version (of the initial request)?

**Comment by mkaply** [comment]
If that was the case, I would think the http would fail and the https would succeed. Odd.

-1 is just our generic download failure.

**Comment by bnk2** [comment]
I retried executing the install function using the debugger as described above and watching the webserver access log (clearing the browser cache before each new attempt).

* HTTPS-URL (HTTP/2.0)
```
<my_client_ip> - - [31/Jul/2019:11:14:54 +0200] "GET /some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi HTTP/2.0" 200 214369 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"
```
* HTTP-URL (HTTP/1.1)
```
<my_client_ip> - - [31/Jul/2019:11:15:27 +0200] "GET /some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi HTTP/1.1" 302 161 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"
<my_client_ip> - - [31/Jul/2019:11:15:27 +0200] "GET /some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi HTTP/2.0" 200 214369 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0"
```

I just tried removing the HTTP/2.0 support on HTTPS...
* HTTPS-URL (HTTP/1.1)
```
<my_client_ip> - - [31/Jul/2019:11:21:17 +0200] "GET /some/path/to/mirrored/addons/english_united_states_dictionary-60.1webext.xpi HTTP/1.1" 200 214136 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0
```
The result was the same: Using HTTPS-Protocol for the install-URL turned out with: `-1 - Download failed`, using the HTTP-Protocol for install-URL, the redirect is followed but the install result is `Installation succeeded`

Anything more I could debug?

**Comment by mkaply** [comment]
Can you go to about:config, right click create boolean.

Create a pref called:

extensions.install.requireBuiltInCerts

and set it to false?

and see what that does?

**Comment by bnk2** [comment]
That made it work.

Is that the way how it should work?

**Comment by mkaply** [comment]
> Is that the way how it should work?

I'm looking into that. I'm guessing that your extensions are being served from an https server that uses your own CA? And it appears that it doesn't chain to a known CA?

**Comment by bnk2** [comment]
Correct, we are using an own CA which is not known in the _Internet_.

But it is imported in the policy section as a _trusted_ one.

It works fine for all web-based applications and other HTTP clients in our Intranet, except downloading (installing) Mozilla extensions with Firefox - and as we saw in the log files, the transport itself seems to be working as well.

**Comment by mkaply** [comment]
So unfortunately addons have a legacy requirement to only allow installing from builtin CAs (or things chained to those CAs.

Of course that seems strange when you can install from http or redirect to an https that isn't chained.

I've opened a bug to remove that requirement.

https://bugzilla.mozilla.org/show_bug.cgi?id=1308251

Are you ok with just using http in this case or should I figure out a way to expose extensions.install.requireBuiltInCerts via policy?

**Comment by bnk2** [comment]
We are fine with the workaround using the redirect of the webserver. It would be nice, this will be fixed in future versions of Firefox.

**Comment by bnk2** [comment]
@mkaply The issue seems to be fixed in Firefox 78.0(.1 ESR).

**Comment by mkaply** [comment]
Awesome.
